pkg/controllers/machine/termination/controller.go

```diff
@@ -0,0 +1,125 @@
+	machineutil "github.com/aws/karpenter-core/pkg/utils/machine"
+)
+
+var _ corecontroller.FinalizingTypedController[*v1alpha5.Machine] = (*Controller)(nil)
+
+// Controller is a Machine Termination controller that triggers deletion of the Node and the
+// CloudProvider Machine through its graceful termination mechanism
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController is a constructor for the Machine Controller
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider) corecontroller.Controller {
+	return corecontroller.Typed[*v1alpha5.Machine](kubeClient, &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	})
+}
+
+func (*Controller) Name() string {
+	return "machine_termination"
+}
+
+func (c *Controller) Reconcile(_ context.Context, _ *v1alpha5.Machine) (reconcile.Result, error) {
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Finalize(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	stored := machine.DeepCopy()
+	if !controllerutil.ContainsFinalizer(machine, v1alpha5.TerminationFinalizer) {
+		return reconcile.Result{}, nil
+	}
+	nodes, err := machineutil.AllNodesForMachine(ctx, c.kubeClient, machine)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if len(nodes) > 0 {
+		for _, node := range nodes {
+			if err := c.kubeClient.Delete(ctx, node); client.IgnoreNotFound(err) != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		return reconcile.Result{}, nil
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("provisioner", machine.Labels[v1alpha5.ProvisionerNameLabelKey], "provider-id", machine.Status.ProviderID))
+	if machine.Status.ProviderID != "" {
+		if err := c.cloudProvider.Delete(ctx, machine); cloudprovider.IgnoreMachineNotFoundError(err) != nil {
+			return reconcile.Result{}, fmt.Errorf("terminating cloudprovider instance, %w", err)
+		}
+	}
+	controllerutil.RemoveFinalizer(machine, v1alpha5.TerminationFinalizer)
+	if !equality.Semantic.DeepEqual(stored, machine) {
+		if err := c.kubeClient.Patch(ctx, machine, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(fmt.Errorf("removing machine termination finalizer, %w", err))
+		}
+		logging.FromContext(ctx).Infof("deleted machine")
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(ctx context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(controllerruntime.
+		NewControllerManagedBy(m).
+		For(&v1alpha5.Machine{}).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
+		Watches(
+			&source.Kind{Type: &v1.Node{}},
+			machineutil.NodeEventHandler(ctx, c.kubeClient),
+			// Watch for node deletion events
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool { return false },
+				UpdateFunc: func(e event.UpdateEvent) bool { return false },
```

Comment:
Do we need to also watch update events since the finalizer logic triggers a node update?

Reply:
Nope, we don't care about the update until the node is actually fully gone; so, Delete() should be sufficient